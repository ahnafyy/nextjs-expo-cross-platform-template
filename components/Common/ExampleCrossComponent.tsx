import React, { forwardRef } from 'react';
import { View, Text, Pressable, ViewProps } from 'react-native';
import tw from 'twrnc';

export type ExampleCrossComponentProps = ViewProps & {
  title?: string;
  variant?: 'primary' | 'secondary' | 'success' | 'warning' | 'error';
  size?: 'sm' | 'md' | 'lg';
  disabled?: boolean;
  loading?: boolean;
  onPress?: () => void;
};

/**
 * ExampleCrossComponent - A demonstration component showing cross-platform patterns
 * 
 * This component demonstrates:
 * - React Native primitives for cross-platform compatibility
 * - TWRNC for unified styling
 * - TypeScript with proper prop typing
 * - forwardRef pattern for ref forwarding
 * - Conditional styling patterns
 * - Accessibility features
 */
const ExampleCrossComponent = forwardRef<View, ExampleCrossComponentProps>(
  ({ 
    title = 'Button', 
    variant = 'primary', 
    size = 'md',
    disabled = false,
    loading = false,
    onPress, 
    style, 
    children, 
    ...props 
  }, ref) => {
    
    // Define base styles using TWRNC
    const baseStyles = tw`rounded-lg items-center justify-center border`;
    
    // Size variations
    const sizeStyles = {
      sm: tw`px-3 py-1.5`,
      md: tw`px-4 py-2.5`,
      lg: tw`px-6 py-3.5`,
    }[size];
    
    // Variant color schemes
    const variantStyles = {
      primary: tw`bg-blue-500 border-blue-500`,
      secondary: tw`bg-gray-500 border-gray-500`,
      success: tw`bg-green-500 border-green-500`,
      warning: tw`bg-yellow-500 border-yellow-500`,
      error: tw`bg-red-500 border-red-500`,
    }[variant];
    
    // Text color for contrast
    const textColorStyles = {
      primary: tw`text-white`,
      secondary: tw`text-white`,
      success: tw`text-white`,
      warning: tw`text-black`,
      error: tw`text-white`,
    }[variant];
    
    // Text size based on component size
    const textSizeStyles = {
      sm: tw`text-sm`,
      md: tw`text-base`,
      lg: tw`text-lg`,
    }[size];
    
    // Disabled state
    const disabledStyles = disabled ? tw`opacity-50` : tw``;
    
    // Loading state (dim the component)
    const loadingStyles = loading ? tw`opacity-75` : tw``;
    
    // Combine all styles
    const finalStyles = [
      baseStyles,
      sizeStyles,
      variantStyles,
      disabledStyles,
      loadingStyles,
      style, // Allow custom styles to override
    ];
    
    const finalTextStyles = [
      textColorStyles,
      textSizeStyles,
      tw`font-medium text-center`,
    ];
    
    return (
      <Pressable
        ref={ref}
        style={finalStyles}
        onPress={disabled || loading ? undefined : onPress}
        accessible={true}
        accessibilityRole="button"
        accessibilityLabel={title}
        accessibilityState={{ disabled: disabled || loading }}
        // Provide visual feedback on press (works on both platforms)
        android_ripple={{ color: 'rgba(255, 255, 255, 0.2)' }}
        style={({ pressed }) => [
          ...finalStyles,
          // iOS-style press feedback
          pressed && tw`opacity-80`,
        ]}
        {...props}
      >
        {loading ? (
          <Text style={[finalTextStyles, tw`animate-pulse`]}>
            Loading...
          </Text>
        ) : (
          <>
            {title && (
              <Text style={finalTextStyles}>
                {title}
              </Text>
            )}
            {children}
          </>
        )}
      </Pressable>
    );
  }
);

ExampleCrossComponent.displayName = 'ExampleCrossComponent';

export default ExampleCrossComponent;