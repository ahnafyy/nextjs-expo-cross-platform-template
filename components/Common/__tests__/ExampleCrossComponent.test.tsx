import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Text } from 'react-native';
import ExampleCrossComponent from '../ExampleCrossComponent';

describe('ExampleCrossComponent', () => {
  // Basic rendering test
  it('renders correctly with default props', () => {
    const { getByText } = render(<ExampleCrossComponent />);
    expect(getByText('Button')).toBeTruthy();
  });

  // Custom title test
  it('renders with custom title', () => {
    const { getByText } = render(
      <ExampleCrossComponent title="Custom Title" />
    );
    expect(getByText('Custom Title')).toBeTruthy();
  });

  // Press event test
  it('handles press events', () => {
    const mockOnPress = jest.fn();
    const { getByRole } = render(
      <ExampleCrossComponent title="Press Me" onPress={mockOnPress} />
    );
    
    const button = getByRole('button');
    fireEvent.press(button);
    
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  // Disabled state test
  it('does not call onPress when disabled', () => {
    const mockOnPress = jest.fn();
    const { getByRole } = render(
      <ExampleCrossComponent 
        title="Disabled Button" 
        disabled 
        onPress={mockOnPress} 
      />
    );
    
    const button = getByRole('button');
    fireEvent.press(button);
    
    expect(mockOnPress).not.toHaveBeenCalled();
  });

  // Loading state test
  it('shows loading text when loading', () => {
    const { getByText } = render(
      <ExampleCrossComponent title="Original" loading />
    );
    
    expect(getByText('Loading...')).toBeTruthy();
  });

  // Loading state does not call onPress
  it('does not call onPress when loading', () => {
    const mockOnPress = jest.fn();
    const { getByRole } = render(
      <ExampleCrossComponent 
        title="Loading Button" 
        loading 
        onPress={mockOnPress} 
      />
    );
    
    const button = getByRole('button');
    fireEvent.press(button);
    
    expect(mockOnPress).not.toHaveBeenCalled();
  });

  // Accessibility test
  it('has correct accessibility properties', () => {
    const { getByRole } = render(
      <ExampleCrossComponent title="Accessible Button" disabled />
    );
    
    const button = getByRole('button');
    
    // Check accessibility role
    expect(button).toBeTruthy();
    
    // Check accessibility state for disabled
    expect(button.props.accessibilityState).toEqual({ disabled: true });
  });

  // Custom children test
  it('renders custom children when no title provided', () => {
    const { getByText } = render(
      <ExampleCrossComponent>
        <Text>Custom Child</Text>
      </ExampleCrossComponent>
    );
    
    expect(getByText('Custom Child')).toBeTruthy();
  });

  // Variant styling test (snapshot)
  it('applies correct variant styles', () => {
    const { toJSON: primaryJSON } = render(
      <ExampleCrossComponent variant="primary" title="Primary" />
    );
    const { toJSON: secondaryJSON } = render(
      <ExampleCrossComponent variant="secondary" title="Secondary" />
    );
    
    // Snapshots help ensure styling consistency
    expect(primaryJSON).toMatchSnapshot('primary-variant');
    expect(secondaryJSON).toMatchSnapshot('secondary-variant');
  });

  // Size styling test
  it('applies correct size styles', () => {
    const { toJSON: smallJSON } = render(
      <ExampleCrossComponent size="sm" title="Small" />
    );
    const { toJSON: largeJSON } = render(
      <ExampleCrossComponent size="lg" title="Large" />
    );
    
    expect(smallJSON).toMatchSnapshot('small-size');
    expect(largeJSON).toMatchSnapshot('large-size');
  });
});