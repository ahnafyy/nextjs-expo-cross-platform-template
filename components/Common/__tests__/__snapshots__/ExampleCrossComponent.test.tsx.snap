// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleCrossComponent applies correct size styles: large-size 1`] = `[Function]`;

exports[`ExampleCrossComponent applies correct size styles: small-size 1`] = `[Function]`;

exports[`ExampleCrossComponent applies correct variant styles: primary-variant 1`] = `[Function]`;

exports[`ExampleCrossComponent applies correct variant styles: secondary-variant 1`] = `[Function]`;
